wifiphisher/common/attacks.py

```diff
@@ -0,0 +1,322 @@
+
+    def iface_up(self):
+
+        """
+        If the interface goes down
+        """
+
+        subprocess.Popen(["sudo ifconfig "+self.iface+" up"], shell=True)
+
+    def check_if_connected(self):
+
+        """
+        This function will keep checking if we
+        are successfully connected to the target AP
+        """
+
+        while 1:
+            conn = self.check_ap() #check if we got an ip from the AP
+            if not conn:
+                time.sleep(1)
+            else :
+                self.done()
+
+    def check_ap(self):
+
+        """
+        Returns the ip that we got or a False if
+        we are not connected to the target AP
+
+        :param self: A WPS_PBC object
+        :type self: WPS_PBC
+        :return: True : Got an IP and connected to the right AP
+                 False : Didn't get an IP or wrong AP ( XD )
+        :rtype: bool
+        """
+
+        s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        while 1:
+            try :
+                ip = socket.inet_ntoa(fcntl.ioctl( #checking if we got an ip from the target AP
+                    s.fileno(),
+                    0x8915,
+                    struct.pack('256s', self.iface[:15])
+                )[20:24])
+                #check the identity of the AP we are connected to
+                conn_ap = (os.popen("iwconfig "+self.iface+" 2> /dev/null | awk -F\Access"
+                                    "'{print $2}'").read()).strip("Point: \n")
+                if conn_ap == self.bssid :
+                    break
+                else :
+                    self.error = "WPS is now connected, but to the wrong AP. AP connected : "+conn_ap
+                    raise AttackError(self.error)
+            except IOError as e:
+                if e.errno == 99:  #if we didn't get the ip return False
+                    return False
+                elif e.errno == 19: #if iface is down for some reasons
+                    self.iface_up()
+                    continue
+                else :
+                    return False
+        return True
+
+
+
+    def done(self):
+
+        """
+        This function will prompt a log to the user,
+        set the self.going to False to terminate the attack
+        """
+
+        with open(self.log_file, "a+") as log_file:
+            log_file.write('[' + T + '*' + W + '] ' + G + "The WPS button has been pressed :" +
+                           C + "You are now connected to "+self.bssid+" the fake AP is still running\n"+
+                           "but the deauthentication attack and the WPS_PBC attack are not running anymore")
+        self.going = False
+
+    def stop(self):
+
+        """
+        The stop function it's different from the done()
+        function because this runs only when 2mins are passed
+        and we are still not connected to the victim's AP.
+        This restart the deauthentication process and change
+        the variable self.going to stop the attack.
```

Comment:
@sophron deauth needs to stop if we want to connect with WPS...

Because of this we can use the same interface for both.

When the deauth pauses to hop channels, we can use that as an opportunity to try connect using WPS